src/mz-debug/src/kubectl_port_forwarder.rs

```diff
@@ -0,0 +1,182 @@
+                                    String::from_utf8_lossy(&output.stderr)
+                                );
+                                error!("{}", retry_err_msg);
+
+                                return RetryResult::RetryableErr(anyhow::anyhow!(retry_err_msg));
+                            }
+                        }
+                        Err(err) => {
+                            return RetryResult::RetryableErr(anyhow::anyhow!(
+                                "Failed to port-forward: {}",
+                                err
+                            ));
+                        }
+                    }
+                    // The kubectl subprocess's future will only resolve on error, thus the
+                    // code here is unreachable. We return RetryResult::Ok to satisfy
+                    // the type checker.
+                    RetryResult::Ok(())
+                }
+            })
+            .await
+        {
+            error!("{}", err);
+        }
+    }
+}
+
+pub async fn create_kubectl_port_forwarder(
+    client: &Client,
+    args: &SelfManagedDebugMode,
+) -> Result<KubectlPortForwarder, anyhow::Error> {
+    for namespace in &args.k8s_namespaces {
+        let services: Api<Service> = Api::namespaced(client.clone(), namespace);
+        let services = services
+            .list(&ListParams::default().labels("materialize.cloud/mz-resource-id"))
+            .await?;
+        // Finds the sql service that contains a port with name "sql"
+        let maybe_port_info = services
+            .iter()
+            .filter_map(|service| {
+                let spec = service.spec.as_ref()?;
+                let service_name = service.metadata.name.as_ref()?;
+                if !service_name.to_lowercase().contains("balancerd") {
+                    return None;
+                }
+                Some((spec, service_name))
+            })
+            .flat_map(|(spec, service_name)| {
+                spec.ports
+                    .iter()
+                    .flatten()
+                    .map(move |port| (port, service_name))
+            })
+            .find_map(|(port_info, service_name)| {
+                if let Some(port_name) = &port_info.name {
+                    // We want to find the external SQL port and not the internal one
+                    if port_name.to_lowercase().contains("pgwire") {
+                        return Some(KubectlPortForwarder {
+                            context: args.k8s_context.clone(),
+                            namespace: namespace.clone(),
+                            service_name: service_name.to_owned(),
+                            target_port: port_info.port,
+                            local_address: args.port_forward_local_address.clone(),
+                            local_port: args.port_forward_local_port,
+                        });
+                    }
+                }
+
+                None
+            });
```

Comment:
There may be multiple Materialize CRs in a single namespace. This will only find and forward one of them.

Reply:
Gah you're right. I'll handle in a follow-up PR, but the way I'm thinking we can handle this is having an optional flag to specify the Materialize CR they want to port forward from?

Reply:
That being said, I recall us discussing how it's uncommon for customers to have multiple Materialize CRs(maybe unless they're doing an upgrade). So this shouldn't be common, right?